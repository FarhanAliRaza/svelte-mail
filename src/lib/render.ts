import type { Component } from 'svelte';
import { render as svelteRender } from 'svelte/server';
import juice from 'juice';
import { convert, type HtmlToTextOptions } from 'html-to-text';
import postcss from 'postcss';
import tailwindcss from 'tailwindcss';
// Try to import @tailwindcss/postcss if available (for Tailwind v4+)
let tailwindPostcss: any;
try {
	// Using dynamic import to handle optional dependency
	tailwindPostcss = require('@tailwindcss/postcss');
} catch (e) {
	// Will use default tailwindcss if @tailwindcss/postcss is not available
}
import type { Config as TailwindConfig } from 'tailwindcss';
import path from 'path';
import fs from 'fs';
// import { marked } from 'marked'; // Import if/when needed

/**
 * Represents the options for the render function.
 */
export interface RenderOptions {
	pretty?: boolean; // Format output HTML
	plainText?: boolean; // Generate plain text version
	tailwindConfig?: string | TailwindConfig; // Path or object for Tailwind configuration
	globalStyles?: string; // Add option for global styles (e.g., fonts)
	htmlToTextOptions?: HtmlToTextOptions; // Options for html-to-text
	juiceOptions?: juice.Options; // Options for juice
}

/**
 * Process HTML with Tailwind CSS.
 * Extracts classes, runs them through Tailwind JIT, and returns the generated CSS.
 *
 * @param html The HTML to process
 * @param tailwindConfig Path to tailwind config or config object
 * @returns The generated CSS
 */
async function processTailwind(
	html: string,
	tailwindConfig?: string | TailwindConfig
): Promise<string> {
	// 1. Extract classes from HTML
	const classRegex = /class="([^"]*)"/g;
	const matches = html.matchAll(classRegex);
	const classSet = new Set<string>();

	for (const match of matches) {
		const classes = match[1].split(/\s+/).filter(Boolean);
		classes.forEach((cls) => classSet.add(cls));
	}

	if (classSet.size === 0) {
		return ''; // No Tailwind classes found
	}

	// 2. Create a temporary CSS file that uses the extracted classes
	const safelist = [...classSet];
	const tempCss = safelist.map((cls) => `.${cls} {}`).join('\n');

	// 3. Load the Tailwind config
	let resolvedConfig: Partial<TailwindConfig> = {};

	if (typeof tailwindConfig === 'string') {
		const configPath = path.resolve(process.cwd(), tailwindConfig);

		if (fs.existsSync(configPath)) {
			try {
				// eslint-disable-next-line @typescript-eslint/no-var-requires
				resolvedConfig = require(configPath);
			} catch (error) {
				console.warn(`Failed to load Tailwind config from ${configPath}:`, error);
			}
		} else {
			console.warn(`Tailwind config file not found at ${configPath}`);
		}
	} else if (tailwindConfig) {
		resolvedConfig = tailwindConfig;
	}

	// 4. Process with Tailwind
	try {
		let postcssPlugins;

		// Use @tailwindcss/postcss if available (for Tailwind v4+), otherwise use direct tailwindcss
		if (tailwindPostcss) {
			postcssPlugins = [
				tailwindPostcss({
					content: [{ raw: html, extension: 'html' }],
					...resolvedConfig
				})
			];
		} else {
			postcssPlugins = [
				tailwindcss({
					content: [{ raw: html, extension: 'html' }],
					...resolvedConfig
				} as any)
			];
		}

		// Process the CSS through postcss
		const result = await postcss(postcssPlugins).process(tempCss, { from: undefined });

		return result.css;
	} catch (error) {
		console.error('Failed to process Tailwind CSS:', error);
		return '';
	}
}

/**
 * Generate plain text version from HTML.
 *
 * @param html The HTML to convert
 * @param options Options for html-to-text
 * @returns The plain text version
 */
function generatePlainText(html: string, options?: HtmlToTextOptions): string {
	const defaultOptions: HtmlToTextOptions = {
		selectors: [
			{ selector: 'head', format: 'skip' },
			{ selector: 'img', format: 'skip' },
			{ selector: 'a', options: { hideLinkHrefIfSameAsText: true } }
		],
		wordwrap: 80
	};

	return convert(html, { ...defaultOptions, ...options });
}

// Remove Svelte/React SSR comments from HTML
function removeSsrComments(html: string): string {
	return html
		.replace(/<!---->/g, '')
		.replace(/<!--\[-->/g, '')
		.replace(/<!--\]-->/g, '');
}

// Remove Svelte/React SSR comments from HTML and other artifacts
function cleanupHtml(html: string): string {
	return html
		.replace(/<!---->/g, '') // Svelte empty comment
		.replace(/<!--\[-->/g, '') // Svelte block start comment
		.replace(/<!--\]-->/g, '') // Svelte block end comment
		.replace(/<!--\[!-->/g, '') // Other potential comment artifacts
		.replace(/\s+onload="this\.__e=event"/g, '') // Remove onload handlers
		.replace(/\s+onerror="this\.__e=event"/g, '') // Remove onerror handlers
		.replace(/\s+class(?=\s|>)/g, '') // Remove empty class attributes (like class or class="")
		.replace(/;\s*undefined/g, ''); // Remove trailing undefined in styles
}

/**
 * Renders a Svelte email component to an HTML string.
 *
 * @param component The root Svelte component to render (e.g., Html).
 * @param options Optional rendering configuration.
 * @param props Props to pass to the component
 * @returns The rendered HTML string or an object with html and text properties if plainText is true.
 */
export async function render<Props extends Record<string, any>>(
	component: Component<any>,
	options?: RenderOptions,
	props?: Props
): Promise<string | { html: string; text: string }> {
	// 1. Render the component using Svelte 5 SSR
	// Svelte 5 returns { head, body }
	const { head, body } = svelteRender(component as any, { props });

	// 2. Combine head and body - Assume the component provides <html> tag
	let combinedHtml = `<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">\n${head}\n${body}`;

	// 3. Process Tailwind CSS if option is provided
	let tailwindGeneratedCss = '';
	if (options?.tailwindConfig) {
		try {
			tailwindGeneratedCss = await processTailwind(combinedHtml, options.tailwindConfig);
		} catch (error) {
			console.warn('Failed to process Tailwind CSS:', error);
		}
	}

	// Combine Tailwind CSS with other global styles (like fonts)
	const allExtraCss = [options?.globalStyles, tailwindGeneratedCss].filter(Boolean).join('\n');

	// 4. Inline CSS using juice
	const juiceOptions: juice.Options = {
		extraCss: allExtraCss, // Use combined global and Tailwind CSS
		removeStyleTags: true, // Remove other style tags potentially added by Svelte
		...options?.juiceOptions
	};

	const inlinedHtml = juice(combinedHtml, juiceOptions);

	// 5. Apply pretty printing if requested
	let finalHtml = inlinedHtml;
	if (options?.pretty) {
		try {
			// Basic pretty printing (consider a more robust library like prettier if needed)
			finalHtml = inlinedHtml
				.replace(/</g, '\n<')
				.replace(/>/g, '>\n')
				.split('\n')
				.map((line) => line.trim())
				.filter(Boolean)
				.join('\n');
		} catch (error) {
			console.warn('Pretty printing failed:', error);
		}
	}

	// Remove Svelte/React SSR comments
	finalHtml = removeSsrComments(finalHtml);

	// Cleanup HTML artifacts
	finalHtml = cleanupHtml(finalHtml);

	// 6. Generate plain text version if requested
	if (options?.plainText) {
		const plainText = generatePlainText(finalHtml, options.htmlToTextOptions);
		return { html: finalHtml, text: plainText };
	}

	return finalHtml;
}

