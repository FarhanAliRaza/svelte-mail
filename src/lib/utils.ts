/**
 * Convert pixels to points for MSO (Outlook) email clients.
 * The conversion ratio is approximately 0.75pt = 1px.
 */
export function pxToPt(px: string | number): string | number {
  const value = typeof px === 'string' ? parseInt(px, 10) : px;
  return Math.floor(value * 0.75);
}

/**
 * Convert a style object to a style string.
 * This is useful for applying styles to HTML elements in email templates.
 */
/**
 * Takes any number of style strings (CSS style attribute format), removes duplicate properties (last one wins),
 * and returns a normalized style string. Supports both camelCase and kebab-case property names.
 * Similar to `clsx`/`cn` but for inline styles.
 *
 * Example:
 *   styleToString('color: red; font-size: 12px;', 'color: blue; margin: 0;')
 *   => 'font-size: 12px; color: blue; margin: 0;'
 */
/**
 * Converts style strings to a normalized, deduplicated, and email-safe style string.
 * - Deduplicates properties (last one wins)
 * - Normalizes property names to kebab-case
 * - Filters out properties known to be unsafe or unsupported in email clients
 * - Escapes values to prevent style injection
 */
export function styleToString(...styles: (string | undefined | null | false)[]): string {
  // List of CSS properties that are generally safe and supported in email clients
  // (This is not exhaustive, but covers most common email-safe properties)
  const EMAIL_SAFE_PROPERTIES = new Set([
    // Box model
    'margin', 'margin-top', 'margin-right', 'margin-bottom', 'margin-left',
    'padding', 'padding-top', 'padding-right', 'padding-bottom', 'padding-left',
    'width', 'height', 'min-width', 'min-height', 'max-width', 'max-height',
    // Typography
    'font', 'font-family', 'font-size', 'font-style', 'font-weight', 'line-height', 'letter-spacing', 'text-align', 'text-decoration', 'text-transform', 'color',
    // Background
    'background', 'background-color',
    // Border
    'border', 'border-top', 'border-right', 'border-bottom', 'border-left',
    'border-width', 'border-style', 'border-color', 'border-radius',
    // Table
    'border-collapse', 'border-spacing',
    // Misc
    'display', 'vertical-align', 'overflow', 'white-space', 'direction',
    // Outlook-specific
    'mso-padding-alt', 'mso-text-raise', 'mso-font-width'
  ]);

  // Escape style values to prevent style injection
  function escapeStyleValue(val: string): string {
    // If dangerous patterns are found, maybe return empty or handle differently?
    if (/expression\s*\(/gi.test(val)) {
      console.warn(`[styleToString] Potentially unsafe value detected (expression): "${val}"`);
      return ''; // Return empty string to effectively discard the rule
    }
    // Basic escaping: remove semicolons, CSS variables, and newlines
    return val
      .replace(/;/g, '') // Remove semicolons within values
      .replace(/--/g, '') // Remove CSS variable syntax
      .replace(/[\r\n]/g, '') // Remove newlines
      .replace(/\\/g, ''); // Remove backslashes
  }

  const styleMap: Record<string, string> = {};

  for (const style of styles) {
    if (!style || typeof style !== 'string') continue;
    const declarations = style.split(';').map(s => s.trim()).filter(Boolean);
    for (const decl of declarations) {
      const colonIndex = decl.indexOf(':');
      if (colonIndex === -1) {
        if (decl.length > 0) {
          console.warn(`[styleToString] Invalid style declaration: "${decl}"`);
        }
        continue;
      }

      let key = decl.substring(0, colonIndex).trim();
      let value = decl.substring(colonIndex + 1).trim();

      // Escape value first
      const escapedValue = escapeStyleValue(value);

      // If escaping resulted in an empty string (due to detecting unsafe pattern), skip this rule
      if (escapedValue === '' && value !== '') {
        continue;
      }

      // Normalize property key
      if (!key.startsWith('mso-') && !key.includes('-')) {
        key = key.replace(/([A-Z])/g, '-$1').toLowerCase();
      }

      // Check if the *normalized* key is safe
      if (!EMAIL_SAFE_PROPERTIES.has(key)) {
        // Check if the *original* key was unsafe (e.g., --custom-prop)
        if (key !== decl.substring(0, colonIndex).trim() || !EMAIL_SAFE_PROPERTIES.has(decl.substring(0, colonIndex).trim())) {
          console.warn(`[styleToString] Unsupported or unsafe CSS property: "${key}" will be ignored.`);
          continue;
        }
      }

      styleMap[key] = escapedValue;
    }
  }

  // Join properties and trim trailing space
  return Object.entries(styleMap)
    .map(([key, value]) => `${key}: ${value};`)
    .join(' ')
    .trim();
}



