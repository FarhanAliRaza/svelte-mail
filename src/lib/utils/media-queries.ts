/**
 * Media query breakpoint definitions
 */
export const breakpoints = {
  mobile: 480,
  tablet: 768,
  desktop: 992
};

/**
 * CSS properties interface for type safety
 */
export interface CSSProperties {
  [key: string]: string | number;
}

/**
 * Media query options for different screen sizes
 */
export interface MediaQueryOptions {
  mobile?: CSSProperties;
  tablet?: CSSProperties;
  desktop?: CSSProperties;
  darkMode?: CSSProperties;
}

/**
 * Convert CSS object to string
 * @param css CSS properties object
 * @returns CSS string
 */
export function convertToCss(css?: CSSProperties): string {
  if (!css) return '';
  
  return Object.entries(css)
    .map(([property, value]) => {
      // Convert camelCase to kebab-case
      const kebabProperty = property.replace(/([A-Z])/g, '-$1').toLowerCase();
      return `${kebabProperty}: ${value};`;
    })
    .join('\n');
}

/**
 * Generate media queries for responsive emails
 * @param selector CSS selector to target
 * @param options Media query options for different screen sizes
 * @returns Complete media query CSS string
 */
export function generateMediaQueries(selector: string, options: MediaQueryOptions): string {
  let mediaQueries = '';
  
  if (options.mobile) {
    mediaQueries += `
@media (max-width: ${breakpoints.mobile}px) {
  ${selector} {
    ${convertToCss(options.mobile)}
  }
}
`;
  }
  
  if (options.tablet) {
    mediaQueries += `
@media (min-width: ${breakpoints.mobile + 1}px) and (max-width: ${breakpoints.tablet}px) {
  ${selector} {
    ${convertToCss(options.tablet)}
  }
}
`;
  }
  
  if (options.desktop) {
    mediaQueries += `
@media (min-width: ${breakpoints.tablet + 1}px) {
  ${selector} {
    ${convertToCss(options.desktop)}
  }
}
`;
  }

  if (options.darkMode) {
    mediaQueries += `
@media (prefers-color-scheme: dark) {
  ${selector} {
    ${convertToCss(options.darkMode)}
  }
}
`;
  }
  
  return mediaQueries;
}

/**
 * Extract custom media queries from inline styles
 * Format: data-mq-mobile="width: 100%;" or data-mq-dark="background-color: #000000;"
 * 
 * @param element HTML element with data-mq attributes
 * @returns Media queries CSS string
 */
export function extractMediaQueriesFromElement(element: Element, uniqueId: string): string {
  const selector = `#${uniqueId}`;
  const options: MediaQueryOptions = {};
  
  // Extract mobile styles
  const mobileStyles = element.getAttribute('data-mq-mobile');
  if (mobileStyles) {
    options.mobile = parseCssString(mobileStyles);
    element.removeAttribute('data-mq-mobile');
  }
  
  // Extract tablet styles
  const tabletStyles = element.getAttribute('data-mq-tablet');
  if (tabletStyles) {
    options.tablet = parseCssString(tabletStyles);
    element.removeAttribute('data-mq-tablet');
  }
  
  // Extract desktop styles
  const desktopStyles = element.getAttribute('data-mq-desktop');
  if (desktopStyles) {
    options.desktop = parseCssString(desktopStyles);
    element.removeAttribute('data-mq-desktop');
  }
  
  // Extract dark mode styles
  const darkModeStyles = element.getAttribute('data-mq-dark');
  if (darkModeStyles) {
    options.darkMode = parseCssString(darkModeStyles);
    element.removeAttribute('data-mq-dark');
  }
  
  return generateMediaQueries(selector, options);
}

/**
 * Parse CSS string to CSS properties object
 * @param cssString CSS string (e.g. "width: 100%; color: red;")
 * @returns CSS properties object
 */
export function parseCssString(cssString: string): CSSProperties {
  const result: CSSProperties = {};
  
  // Split the string by semicolons and process each declaration
  const declarations = cssString.split(';').filter(Boolean);
  
  for (const declaration of declarations) {
    // Split the declaration into property and value
    const [property, value] = declaration.split(':').map(part => part.trim());
    
    if (property && value) {
      // Convert kebab-case to camelCase
      const camelProperty = property.replace(/-([a-z])/g, (_, char) => char.toUpperCase());
      result[camelProperty] = value;
    }
  }
  
  return result;
} 