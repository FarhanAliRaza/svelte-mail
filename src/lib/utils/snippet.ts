import { createRawSnippet, type Snippet } from 'svelte';

// Correctly use createRawSnippet for static text
export const createSnippet = (text: string): Snippet => {
	return createRawSnippet(() => {
		// Factory function (takes no args here)
		return {
			render: () => text // Render method returns the captured text
			// setup is optional for basic SSR mocking
		};
	});
};
