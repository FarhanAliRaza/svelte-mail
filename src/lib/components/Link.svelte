<script lang="ts">
	import type { Snippet } from 'svelte';
	import type { HTMLAttributes } from 'svelte/elements';
	import { styleToString } from '../utils';

	interface Props extends Omit<HTMLAttributes<HTMLAnchorElement>, 'style' | 'target'> {
		href: string;
		target?: string;
		style?: string | Record<string, string | number>;
		children?: Snippet;
		[key: string]: any;
	}

	let { href, target = '_blank', style = {}, children, ...rest }: Props = $props();

	const defaultStyle = 'color: #067df7; text-decoration-line: none;';

	function normalizeStyleProp(styleProp: Props['style']): string | undefined {
		if (typeof styleProp === 'string') {
			return styleProp;
		} else if (typeof styleProp === 'object' && styleProp !== null) {
			return Object.entries(styleProp)
				.map(([key, value]) => `${key}:${value};`)
				.join(' ');
		}
		return undefined;
	}

	const normalizedPassedStyle = normalizeStyleProp(style);

	const combinedStyle = styleToString(defaultStyle, normalizedPassedStyle);
</script>

<a {href} {target} style={combinedStyle} {...rest}>
	{#if children}
		{@render children()}
	{/if}
</a>
