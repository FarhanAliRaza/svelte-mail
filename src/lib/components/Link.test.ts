// @vitest-environment node
import { describe, it, expect } from 'vitest';
import Link from './Link.svelte';
import { render } from '../render';
import { createSnippet } from '../utils/snippet';

describe('Link component (SSR)', () => {
	it('renders children correctly', async () => {
		const testMessage = 'Test message';
		const result = await render(Link, undefined, {
			href: 'https://example.com',
			children: createSnippet(testMessage)
		});

		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('href="https://example.com"');
		expect(html).toMatchSnapshot();
	});

	it('passes style and other props correctly', async () => {
		const style = { color: 'red' };
		const result = await render(Link, undefined, {
			href: 'https://example.com',
			style,
			'data-testid': 'link-test',
			children: createSnippet('Test')
		});

		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('color: red;');
		expect(html).toContain('data-testid="link-test"');
		expect(html).toMatchSnapshot();
	});

	it('opens in a new tab by default', async () => {
		const result = await render(Link, undefined, {
			href: 'https://example.com',
			children: createSnippet('Test')
		});

		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('target="_blank"');
		expect(html).toMatchSnapshot();
	});

	it('renders correctly with all props', async () => {
		const result = await render(Link, undefined, {
			href: 'https://example.com',
			style: { color: 'green' },
			target: '_self',
			children: createSnippet('Example')
		});

		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('href="https://example.com"');
		expect(html).toContain('color: green;');
		expect(html).toContain('target="_self"');
		expect(html).toMatchSnapshot();
	});
});
