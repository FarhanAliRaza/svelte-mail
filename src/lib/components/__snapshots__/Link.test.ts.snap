// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Link component (SSR) > opens in a new tab by default 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" style="color: #067df7;">Test</a>"
`;

exports[`Link component (SSR) > passes style and other props correctly 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" style="color: red;" data-testid="link-test">Test</a>"
`;

exports[`Link component (SSR) > renders children correctly 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" style="color: #067df7;">Test message</a>"
`;

exports[`Link component (SSR) > renders correctly with all props 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_self" style="color: green;">Example</a>"
`;
